File: src/main/java/com/rokucraft/rokubot/command/commands/IssueCommand.java
package com.rokucraft.rokubot.command.commands;

import com.rokucraft.rokubot.command.AutoCompletable;
import com.rokucraft.rokubot.command.SlashCommand;
import com.rokucraft.rokubot.util.EmbedUtil;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.Command.Choice;
import net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.kohsuke.github.*;

import java.io.IOException;
import java.util.List;

import static com.rokucraft.rokubot.Constants.ISSUE_CLOSED_COLOR;
import static com.rokucraft.rokubot.Constants.ISSUE_OPEN_COLOR;

public class IssueCommand implements SlashCommand, AutoCompletable {
    private final @NonNull CommandData data;
    private final @NonNull GitHub github;
    private final @Nullable List<GHRepository> repositoryCache;
    private final @Nullable String defaultRepoName;
    public static final String ICON_ISSUE_OPEN =
            "https://cdn.discordapp.com/attachments/786216721065050112/787721554992824360/issue-opened72px.png";
    public static final String ICON_ISSUE_CLOSED =
            "https://cdn.discordapp.com/attachments/786216721065050112/787721551285059614/issue-closed72px.png";

    public IssueCommand(
            @NonNull GitHub github,
            @Nullable List<GHRepository> repositoryCache,
            @Nullable String defaultRepoName
    ) {
        this.github = github;
        this.repositoryCache = repositoryCache;
        this.defaultRepoName = defaultRepoName;
        this.data = Commands.slash("issue", "Preview an issue")
                .setDefaultPermissions(DefaultMemberPermissions.DISABLED)
                .addOptions(
                        new OptionData(OptionType.INTEGER, "number", "The issue number")
                                .setRequiredRange(1, Integer.MAX_VALUE)
                )
                .addOption(OptionType.STRING, "repo", "The repository the issue belongs to", false,true);
    }

    @Override
    public void execute(SlashCommandInteractionEvent event) {
        try {
            Integer number = event.getOption("number", OptionMapping::getAsInt);
            if (number == null) {
                throw new IllegalArgumentException("You must enter a number");
            }
            String repo = event.getOption("repo", this.defaultRepoName, OptionMapping::getAsString);
            GHIssue issue = this.github.getRepository(repo).getIssue(number);
            event.replyEmbeds(createIssueEmbed(issue)).queue();
        } catch (IOException e) {
            event.replyEmbeds(EmbedUtil.createErrorEmbed("Unable to find issue")).setEphemeral(true).queue();
        } catch (IllegalArgumentException e) {
            event.replyEmbeds(EmbedUtil.createErrorEmbed(e.getMessage())).setEphemeral(true).queue();
        }
    }

    @Override
    public void autoComplete(CommandAutoCompleteInteractionEvent event) {
        String query = event.getFocusedOption().getValue().toLowerCase();
        if (repositoryCache == null) {
            event.replyChoices().queue();
            return;
        }
        event.replyChoices(
                this.repositoryCache
                        .stream()
                        .filter(GHRepository::hasIssues)
                        .filter(r -> r.getName().toLowerCase().contains(query))
                        .limit(25)
                        .map(r -> new Choice(r.getName(), r.getFullName()))
                        .toList()
        ).queue();
    }

    @NonNull
    private MessageEmbed createIssueEmbed(@NonNull GHIssue issue) throws IOException {
        boolean open = issue.getState() == GHIssueState.OPEN;
        GHUser author = issue.getUser();
        String authorName = (author.getName() != null) ? author.getName() : author.getLogin();
        String issueBody = issue.getBody()
                .replaceAll("(?s)<!--.*?-->", "") // Removes HTML comments
                .replaceAll("###(.*)", "**$1**") //Makes Markdown titles bold
                .trim();
        if (issueBody.length() > MessageEmbed.DESCRIPTION_MAX_LENGTH) {
            issueBody = issueBody.substring(0, MessageEmbed.DESCRIPTION_MAX_LENGTH - 1).concat("…");
        }
        return new EmbedBuilder()
                .setColor(open ? ISSUE_OPEN_COLOR : ISSUE_CLOSED_COLOR)
                .setThumbnail(open ? ICON_ISSUE_OPEN : ICON_ISSUE_CLOSED)
                .setAuthor(authorName, author.getHtmlUrl().toString(), author.getAvatarUrl())
                .setTitle(issue.getTitle(), issue.getHtmlUrl().toString())
                .setDescription(issueBody)
                .setTimestamp(issue.getCreatedAt().toInstant())
                .build();
    }

    @Override @NonNull
    public CommandData getData(Guild guild) {
        return this.data;
    }
}

File: src/main/kotlin/com/rokucraft/rokubot/command/commands/IssueCommand.kt
package com.rokucraft.rokubot.command.commands

import com.rokucraft.rokubot.command.AbstractCommand
import com.rokucraft.rokubot.github.RepositoryCache
import com.rokucraft.rokubot.util.ColorConstants.ISSUE_CLOSED_COLOR
import com.rokucraft.rokubot.util.ColorConstants.ISSUE_OPEN_COLOR
import com.rokucraft.rokubot.util.EmbedUtil.createErrorEmbed
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.Command
import net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import org.kohsuke.github.GHIssue
import org.kohsuke.github.GHIssueState
import org.kohsuke.github.GitHub
import java.io.IOException
import java.util.*

class IssueCommand(
    private val github: GitHub,
    private val repositoryCache: RepositoryCache,
    private val defaultRepoName: String?
) : AbstractCommand() {

    override val data = Commands.slash("issue", "Preview an issue")
        .setDefaultPermissions(DefaultMemberPermissions.DISABLED)
        .addOptions(OptionData(OptionType.INTEGER, "number", "The issue number").setMinValue(1))
        .addOption(OptionType.STRING, "repo", "The repository the issue belongs to", false, true)

    override fun execute(event: SlashCommandInteractionEvent) {
        try {
            val number = event.getOption("number") { it.asInt }
                ?: throw IllegalArgumentException("You must enter a number")
            val repo = event.getOption("repo", defaultRepoName) { it.asString }
            val issue = github.getRepository(repo).getIssue(number)
            event.replyEmbeds(createIssueEmbed(issue)).queue()
        } catch (e: IOException) {
            event.replyEmbeds(createErrorEmbed("Unable to find issue")).setEphemeral(true).queue()
        } catch (e: IllegalArgumentException) {
            event.replyEmbeds(createErrorEmbed(e.message ?: "Something went wrong")).setEphemeral(true).queue()
        }
    }

    override fun autoComplete(event: CommandAutoCompleteInteractionEvent) {
        val query = event.focusedOption.value.lowercase()
        event.replyChoices(
            repositoryCache.repositories
                .filter { it.hasIssues() }
                .filter { it.name.lowercase().contains(query) }
                .take(25)
                .map { Command.Choice(it.name, it.fullName) }
        ).queue()
    }
}

@Throws(IOException::class)
private fun createIssueEmbed(issue: GHIssue): MessageEmbed {
    val open = issue.state == GHIssueState.OPEN
    val author = issue.user
    val authorName = if (author.name != null) author.name else author.login
    var issueBody = issue.body
        .replace(Regex("(?s)<!--.*?-->"), "") // Removes HTML comments
        .replace(Regex("###(.*)"), "**$1**") //Makes Markdown titles bold
        .trim()
    if (issueBody.length > MessageEmbed.DESCRIPTION_MAX_LENGTH) {
        issueBody = issueBody.substring(0, MessageEmbed.DESCRIPTION_MAX_LENGTH - 1) + "…"
    }
    return EmbedBuilder()
        .setColor(if (open) ISSUE_OPEN_COLOR else ISSUE_CLOSED_COLOR)
        .setThumbnail(if (open) ICON_ISSUE_OPEN else ICON_ISSUE_CLOSED)
        .setAuthor(authorName, author.htmlUrl.toString(), author.avatarUrl)
        .setTitle(issue.title, issue.htmlUrl.toString())
        .setDescription(issueBody)
        .setTimestamp(issue.createdAt.toInstant())
        .build()
}

private const val ICON_ISSUE_OPEN =
    "https://cdn.discordapp.com/attachments/786216721065050112/787721554992824360/issue-opened72px.png"
private const val ICON_ISSUE_CLOSED =
    "https://cdn.discordapp.com/attachments/786216721065050112/787721551285059614/issue-closed72px.png"
